solvation_analysis/analysis.py

```diff
@@ -0,0 +1,321 @@
+            # TODO: specify start and stop args
+            rdf.run(**self.rdf_run_kwargs)
+            bins, data = rdf.results.bins, rdf.results.rdf
+            self.rdf_data[name] = (data, bins)
+            # generate and save plots
+            if name not in self.radii.keys():
+                self.radii[name] = self.kernel(bins, data, **self.kernel_kwargs)
+            fig, ax = self._plot_solvation_radius(bins, data, self.radii[name])
+            ax.set_title(f"Solvation distance of {name}")
+            self.rdf_plots[name] = fig, ax
+        assert self.solvents.keys() == self.radii.keys(), "Radii missing."
+
+    def _single_frame(self):
+        """
+        This function finds the solvation shells of each solute at a given time step.
+        """
+        # initialize empty arrays
+        all_pairs_list = []
+        all_dist_list = []
+        all_tags_list = []
+        # loop to find solvated atoms of each type
+        for name, solvent in self.solvents.items():
+            pairs, dist = capped_distance(
+                self.solute.positions,
+                solvent.positions,
+                self.radii[name],
+                box=self.u.dimensions,
+            )
+            # replace local ids with absolute ids
+            pairs[:, 1] = solvent.ids[[pairs[:, 1]]]  # TODO: ids vs ix?
+            # extend
+            all_pairs_list.append(pairs)
+            all_dist_list.append(dist)
+            all_tags_list.append(np.full(len(dist), name))  # creating a name array
+        all_pairs = np.concatenate(all_pairs_list, dtype=int)
+        all_dist = np.concatenate(all_dist_list)
+        all_tags = np.concatenate(all_tags_list)
+        frame_length = len(all_pairs)
+        all_frames = np.full(frame_length, self._ts.frame)
+        # put the data into a data frame
+        all_resid = self.u.atoms[all_pairs[:, 1]].resids
+        solvation_data_np = np.column_stack(
+            (all_frames, all_pairs[:, 0], all_pairs[:, 1], all_dist, all_tags, all_resid)
+        )
+        self.solvation_frames.append(solvation_data_np)
+
+    def _conclude(self):
+        """
+        Instantiates the SolvationData class and several analysis classes.
+        """
+        solvation_data_np = np.vstack(self.solvation_frames)
+        solvation_data_df = pd.DataFrame(
+            solvation_data_np,
+            columns=["frame", "solvated_atom", "atom_id", "dist", "res_name", "res_id"]
+        )
+        # clean up solvation_data df
+        for column in ["frame", "solvated_atom", "atom_id", "dist", "res_id"]:
+            solvation_data_df[column] = pd.to_numeric(solvation_data_df[column])
+        solvation_data_dup = solvation_data_df.sort_values(["frame", "solvated_atom", "dist"])
+        solvation_data = solvation_data_dup.drop_duplicates(["frame", "solvated_atom", "res_id"])
+        self.solvation_data_dup = solvation_data_dup.set_index(["frame", "solvated_atom", "atom_id"])
```

Comment:
what is `solvation_data_dup`? needs to be documented.

Reply:
I'll add documentation here. In `solvation_data`, I remove atoms from duplicated residues, in `solvation_data_dup`, I keep all atoms even if they are from the same residue. For most analyses, duplicated atoms should be thrown out, but for distinguishing monodentate / bidentate solvents, the duplicate atoms will need to be held onto.

Maybe a memory cheaper solution would be to just keep the duplicate atoms and then add them back in when needed. That would save me from keeping an extra copy of the data frame.

Reply:
Got it. I clarified this in the docs.